mirrord-operator/templates/config-map.yaml

```diff
@@ -8,7 +8,9 @@ metadata:
 data:
   agent-config.yaml: |-
     {{- if .Values.agent.image }}
-    image: "{{ .Values.agent.image }}"
+    image:
```

Comment:
this breaks API. need to support supplying image string.

Reply:
Makes sense, how do I do that though?

Check for `{{ if .Values.agent.image.registry }}`, and if it's filled we know we're in the new config type?

Reply:
I think you can check for a type ( since you might have also image.tag set only) - if map do new if string do old

Reply:
also need to release operator and bump version for this helm chart to work

Reply:
```yaml
    {{- if .Values.agent.image }}
    {{- if kindIs "string" .Values.agent.image }}
    image: "{{ .Values.agent.image }}"
    {{- else }}
    image:
      {{- if .Values.agent.image.registry }}
      registry: "{{ .Values.agent.image.registry }}"
      {{- end }}
      {{- if .Values.agent.image.tag}}
      tag: "{{ .Values.agent.image.tag }}"
      {{- end }}
    {{- end }}
    {{- end }}
```

This works, but we get this warning when using the `image.registry` and `image.tag` type:

```sh
coalesce.go:289: warning: destination for mirrord-operator.agent.image is a table. Ignoring non-table value ()
```

Reply:
I think helm doesn't support sum types (an `Either`-thingy in our case)?

Reply:
Changed it to be an `if &&`, but problem still persist.

If you do `image: {}`, then it shows **no** warning when setting it as

```yaml
image:
  registry: "foo"
```

But we get a warning if we do:

```yaml
image: "foo"
```

And if we set `image: ""`, then the warnings are reversed (**yes** warning for explicit, **no** warning for just `image: "foo"`).

Reply:
https://github.com/helm/helm/issues/8026

I managed to get it to work without warnings, but then we can't have defaults lol.

Reply:
All of these are working:

```yaml
agent:
  image: ""
```

```yaml
agent:
  image:
    registry: "foo"
    tag: "bar"
```

```yaml
agent:
  image:
    # registry: "foo"
    tag: "bar"
```

```yaml
agent:
  image:
    registry: "foo"
    # tag: "bar"
```

Doesn't work if you set nothing:

```yaml
agent:
  # image: "foo"
```

```sh
Error: template: mirrord-operator/templates/config-map.yaml:10:23: executing "mirrord-operator/templates/config-map.yaml" at <.Values.agent.image>: nil pointer evaluating interface {}.image
helm.go:84: [debug] template: mirrord-operator/templates/config-map.yaml:10:23: executing "mirrord-operator/templates/config-map.yaml" at <.Values.agent.image>: nil pointer evaluating interface {}.image
```

Is this what we want?

Reply:
just need else if (only if agent.image is set)

Reply:
Sorry, I don't follow.

```yaml
    {{- if and (.Values.agent.image) (kindIs "map" .Values.agent.image) }}
    image:
      {{- if .Values.agent.image.registry }}
      registry: "{{ .Values.agent.image.registry }}"
      {{- end }}
      {{- if .Values.agent.image.tag}}
      tag: "{{ .Values.agent.image.tag }}"
      {{- end }}
    {{- else }}     <-- turn this into `else if`?
    image: "{{ .Values.agent.image }}"
    {{- end }}

```

Reply:
Added an `if` for `.Values.agent`, so now we only look into the `agent` options if there is something there (no null access).